### PR TITLE
DietPi-Services 2.0

### PR DESCRIPTION
**Pull request by Fourdee**
## DietPi-Services v2.0

- REF: https://www.freedesktop.org/software/systemd/man/systemd.exec.html#Scheduling
- Migrates process tool into DietPi-Services
- No dupe service code in 2 scripts
- Uses Systemd droppings (:wink:) to apply nice/affinity/scheduler etc, instead of applying to each process afterwards
- Should speed up boot time and service start time | testing required
- :u6307: Should apply process settings to child processes via systemd

#### ToDo:
- :u6307: Code cleanup
- :u6307: Patch dietpi-services.desktop entry to existing installs
- :u6307: add IO scheduler options in GUI?
- :u6307: add support for masking services in GUI
- :u6307: sysvinit support?
- :u6307: Changelog

---

Dev:
```
wget https://raw.githubusercontent.com/MichaIng/DietPi/49e95bd8218b3bf8c39733cf968086806ec403ed/dietpi/dietpi-services -O /DietPi/dietpi/dietpi-services
```

### CODE REVIEW

**Comment by Fourdee** [comment]
Preview:

image

**Comment by MichaIng** [comment]
@Fourdee 
Would be btw great, if it was possible to skip attribute applying for processes. E.g. nice, affinity etc. is only applied if user actively set a value, otherwise system defaults are left, process(es) stay untouched. Then it makes sense as well to have an option to revert attributes to system defaults.

This should speed up the loop quite much, since most users will only apply certain attributes to certain processes/services.

It requires to not pre-generate the arrays for all processes but only create a certain entry when user actively changes/sets it. Then as well only save the existing array entries to settings file. On "revert to default" option unset the related array index.

What do you think?

**Comment by Fourdee** [comment]
@MichaIng 

>  is only applied if user actively set a value, otherwise system defaults are left, process(es) stay untouched

Yep, that is currently how it is. Unless the user edits a service, which triggers ```Save_Process_Tool``` systemd drop in file creation, no process tool is applied.

Only when the file exists, does systemd then apply it.

The init section is required to generate our arrays, based on untouched defaults:
```
						aCPU_NICE[$i]=0
						aCPU_AFFINITY[$i]="0-$(( $G_HW_CPU_CORES - 1 ))"

						aCPU_SCHEDULE_POLICY[$i]='other'
						aCPU_SCHEDULE_PRIORITY[$i]=0
						aIO_SCHEDULE_POLICY[$i]='none'
						aIO_PRIORITY[$i]='0'
```
However, none of the above gets applied, unless user edits service to create the systemd drop in.

**Comment by Fourdee** [comment]
@MichaIng 

> Then it makes sense as well to have an option to revert attributes to system defaults.

:+1:

**Comment by MichaIng** [comment]
@Fourdee 
> Yep, that is currently how it is.

Ah of course, when being applied via systemd drop-in then my old thoughts are somewhat obsolete. Perfect. Btw if I am not mistaken this should work for sysvinit services as well since, since systemd generates systemd units from them. But not sure if drop-ins from /etc/systemd/system are respected for those, needs testing.
```
2019-05-16 14:14:40 root@micha:/tmp# cat /run/systemd/generator.late/dropbear.service
# Automatically generated by systemd-sysv-generator

[Unit]
Documentation=man:systemd-sysv-generator(8)
SourcePath=/etc/init.d/dropbear
Description=LSB: Lightweight SSH server
Before=multi-user.target
Before=multi-user.target
Before=multi-user.target
Before=graphical.target
After=remote-fs.target

[Service]
Type=forking
Restart=no
TimeoutSec=5min
IgnoreSIGPIPE=no
KillMode=process
GuessMainPID=no
RemainAfterExit=yes
SuccessExitStatus=5 6
ExecStart=/etc/init.d/dropbear start
ExecStop=/etc/init.d/dropbear stop
```

**Comment by Fourdee** [comment]
@MichaIng 
Hmm, tried with both
```
/lib/systemd/system/pihole-FTL.d/dietpi-process_tool.conf
/etc/init.d/pihole-FTL.d/dietpi-process_tool.conf
```

No effect.

```
root@DietPi:~# systemd-delta --type=extended
[EXTENDED]   /lib/systemd/system/rc-local.service → /lib/systemd/system/rc-local.service.d/debian.conf
[EXTENDED]   /lib/systemd/system/systemd-timesyncd.service → /lib/systemd/system/systemd-timesyncd.service.d/disable-with-time-daemon.conf
[EXTENDED]   /lib/systemd/system/cron.service → /lib/systemd/system/cron.service.d/dietpi-process_tool.conf
[EXTENDED]   /lib/systemd/system/systemd-resolved.service → /lib/systemd/system/systemd-resolved.service.d/resolvconf.conf
```

**Comment by MichaIng** [comment]
@Fourdee 
Usually I would try `/etc/systemd/system/pihole-FTL.d/` since this overrides `/lib/systemd/system/` settings at least.

**Comment by Fourdee** [comment]
@MichaIng 

Worth a shot, but no joy :frowning_face: 
```
root@DietPi:~# ls -lha /etc/systemd/system/pihole-FTL.d
total 12K
drwxr-xr-x  2 root root 4.0K May 16 14:26 .
drwxr-xr-x 12 root root 4.0K May 16 14:26 ..
-rw-r--r--  1 root root  140 May 16 14:26 dietpi-process_tool.conf
root@DietPi:~# systemctl daemon-reload
root@DietPi:~# systemctl restart pihole-FTL
root@DietPi:~# systemd-delta --type=extended
[EXTENDED]   /lib/systemd/system/cron.service → /lib/systemd/system/cron.service.d/dietpi-process_tool.conf
[EXTENDED]   /lib/systemd/system/systemd-resolved.service → /lib/systemd/system/systemd-resolved.service.d/resolvconf.conf
[EXTENDED]   /lib/systemd/system/systemd-timesyncd.service → /lib/systemd/system/systemd-timesyncd.service.d/disable-with-time-daemon.conf
[EXTENDED]   /lib/systemd/system/rc-local.service → /lib/systemd/system/rc-local.service.d/debian.conf
```

Workaround our end, replace with custom systemd service, full rewrite or simply launch the init.d script? Might cause issues with pihole updates.

EDIT, OR:
use dietpi-process_tool for old init.d backwards compat?

mmm

**Comment by Fourdee** [comment]
Notes, following are ```init.d```:
- Emonhub
- Pi-Hole
- Haproxy
- couchpotato

**Comment by Fourdee** [comment]
@MichaIng 

Seems the PHP repo is causing some issues with deps:
```
libssl-dev : Depends: libssl1.1 (= 1.1.1b-1+0~20190228092419.10+stretch~1.gbp729d89) but 1.1.0j-1~deb9u1 is to be installed

root@DietPi:~# apt-cache policy libssl-dev
libssl-dev:
  Installed: (none)
  Candidate: 1.1.1b-1+0~20190228092419.10+stretch~1.gbp729d89
  Version table:
     1.1.1b-1+0~20190228092419.10+stretch~1.gbp729d89 500
        500 https://packages.sury.org/php stretch/main arm64 Packages
     1.1.0j-1~deb9u1 500
        500 https://deb.debian.org/debian stretch/main arm64 Packages
        500 https://deb.debian.org/debian-security stretch/updates/main arm64 Packages

---
root@DietPi:~# apt-cache policy libssl1.1
libssl1.1:
  Installed: 1.1.0j-1~deb9u1
  Candidate: 1.1.0j-1~deb9u1
  Version table:
     1.1.1b-1+0~20190228092419.10+stretch~1.gbp729d89 -1
        500 https://packages.sury.org/php stretch/main arm64 Packages
 *** 1.1.0j-1~deb9u1 500
        500 https://deb.debian.org/debian stretch/main arm64 Packages
        500 https://deb.debian.org/debian-security stretch/updates/main arm64 Packages
        100 /var/lib/dpkg/status


```

:u6307: Set lower priority for php repo? Fixes above issue.
```
cat << _EOF_ > /etc/apt/preferences.d/sury-php-pin
Package: *
Pin: origin packages.sury.org
Pin-Priority: 100
_EOF_

```

:u5408: Breaks php7.3 install
```
root@DietPi:~# apt-get install php7.3-fpm php7.3-curl php7.3-gd php7.3-mbstring php7.3-xml php7.3-zip php7.3-sqlite3
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 php7.3-fpm : Depends: php7.3-cli but it is not going to be installed
              Depends: libpcre2-8-0 (>= 10.32) but 10.22-3 is to be installed
 php7.3-zip : Depends: libzip4 (>= 1.3.2) but 1.1.2-1.1+b1 is to be installed
E: Unable to correct problems, you have held broken packages.
```

---

:u5408: ```php7.3-apcu``` does not exist.
```
E: Package 'php7.3-apcu' has no installation candidate                                             
                                                                                                                                                                                       

--------------------------------------------------------------------

[FAILED] DietPi-Software | Unable to continue, DietPi-Software will now terminate.

root@DietPi:~# apt-cache policy php7.3
php7.3:
  Installed: (none)
  Candidate: 7.3.5-1+0~20190503093827.38+stretch~1.gbp60a41b
  Version table:
     7.3.5-1+0~20190503093827.38+stretch~1.gbp60a41b 100
        100 https://packages.sury.org/php stretch/main arm64 Packages
root@DietPi:~# apt-cache policy php7.3-apcu
php7.3-apcu:
  Installed: (none)
  Candidate: (none)
  Version table:
```
https://packages.sury.org/php/pool/main/p/php-apcu/ 5.1

---

:u6307: Fixed all with:
```
cat << _EOF_ > /etc/apt/preferences.d/sury-php-pin
Package: libssl-dev
Pin: origin packages.sury.org
Pin-Priority: 100
_EOF_

#dietpi-software, php-apcu
```

@MichaIng 

Ah, see you fixed some of this already on ```dev``` :+1: 

Will send commit to fix the ```php-apcu```

**Comment by MichaIng** [comment]
@Fourdee 
~It should be the other way round: Not installing `libssl-dev` from sury.org but stay with the Debian/Raspbian version. Should be already the case with v6.24.~
**EDIT: Ah interpreted your solution wrong. Jep that basically does the same the our online solution below. However it is required then to actively install any dependency package from the repo, if an insufficient version is available on Debian/Raspbian repo. Also e.g. `phpmyadmin` would install PHP7.0 since it depends on the PHP meta packages (php-cli, php-...) which are then pulled from Stretch + their php7.0-cli, ...
So I suggest we go/stay with the below solution (which is already live in v6.24) to pin down all libssl-related packages only, which are the only ones known to cause problems and never required.**

Could you paste: `cat /etc/apt/preferences.d/dietpi-openssl`

Reason to not use sury.org version is that is breaks Lighttpd on Stretch. And there is no single package on sury.org that does not work (also dependency wise) with the Stretch version (1.1.0 vs. 1.1.1).

```
echo -e '# libssl1.1 from sury.org breaks Lighttpd install
Package: openssl libssl*\nPin: origin packages.sury.org\nPin-Priority: -1' > /etc/apt/preferences.d/dietpi-openssl
```
Solves all this, pinning down libssl-docs and libssl1.1-dbgsym as well from the repo.

**Comment by Fourdee** [comment]
:u6307: Wrapper to use existing ```/etc/init.d``` but allow systemd drop ins
```
cat << _EOF_ > /lib/systemd/system/pihole-FTL.service
[Unit]
Description=Pi-hole (DietPi)

[Service]
Type=oneshot
RemainAfterExit=yes
ExecStop=/bin/bash -c '/etc/init.d/pihole-FTL stop'
ExecStart=/bin/bash -c '/etc/init.d/pihole-FTL start'

[Install]
WantedBy=multi-user.target
_EOF_
systemctl daemon-reload
systemctl restart pihole-FTL
```

**Comment by MichaIng** [comment]
@Fourdee 
Perhaps we should contribute a real systemd unit to Pi-hole middle-term that does not invoke the sysvinit service 😉. Still going with sysvinit is simply out of time IMO. At least one should have both but sysvinit as fallback only.

> ExecStop=

Doesn't this require `RemainAfterExit=true` in case of oneshot? Otherwise `Type=forking`?

**Comment by MichaIng** [comment]
@Fourdee 
> php7.3-apcu does not exist.

As long as no dependency issues occur, APT selects the correct packages automatically due to: `Provides: php-user-cache, php7.3-apcu`

```
root@DietPi:~# apt install php7.3-apcu php7.3-redis
Reading package lists... Done
Building dependency tree
Reading state information... Done
Note, selecting 'php-apcu' instead of 'php7.3-apcu'
Note, selecting 'php-redis' instead of 'php7.3-redis'
```

So I prefer to stray with explicitly select the 7.3/$PHP_NAME version to prevent accidental wrong version installs.

**Comment by Fourdee** [comment]
EDIT: https://github.com/MichaIng/DietPi/pull/2786#issuecomment-493758912

**Comment by Fourdee** [comment]
@MichaIng 

> So I prefer to stray with explicitly select the 7.3/$PHP_NAME version to prevent accidental wrong version installs.

Also seems php-apcu dev has stopped at version 5:
https://packages.sury.org/php/pool/main/p/php-apcu/
https://packages.debian.org/sid/php-apcu

**Comment by Fourdee** [comment]
Backwards cap for init.d, create systemd wrapper on fly test:
- :u6307: Pi-hole
- :u6307: haproxy

**Comment by MichaIng** [comment]
@Fourdee 
> E: Package 'php7.3-apcu' has no installation candidate

Very strange, tested up and down on all systems and as well no user reports about this so far. This only showed up when there is a dependency issue that prevents install anyway.

Can you paste:
```
apt show php-apcu
apt install php-apcu # Just to check if it actually would install or not
```

> Also seems php-apcu dev has stopped at version 5

This version number is not related to PHP at all, similar to php-redis: https://packages.debian.org/buster/php-redis
- Those packages are developed outside the PHP core modules, which is also the reason why they do not contain the PHP version string.
- Basically they work with all PHP versions, as long as the APT dependencies are set correctly, which is why e.g. you will find all different PHP version sub dirs in /etc/php, each with the php-apcu setting inside.
- The apt show command should also show:
`Provides: php-user-cache, php5.6-apcu, php7.0-apcu, php7.1-apcu, php7.2-apcu, php7.3-apcu`

**Comment by Fourdee** [comment]
@MichaIng 

Yep, installs fine:
```
root@DietPi:~# apt show php-apcu
Package: php-apcu                                                                                                                                            
Version: 5.1.17+4.0.11-1+0~20190217111312.9+stretch~1.gbp192528                                                                                              
Priority: optional                                                                                                                                           
Section: php                                                                                                                                                 
Maintainer: Debian PHP PECL Maintainers <team+php-pecl@tracker.debian.org>                                                                                   
Installed-Size: 781 kB                                                                                                                                       
Provides: php-user-cache, php5.6-apcu, php7.0-apcu, php7.1-apcu, php7.2-apcu, php7.3-apcu                                                                    
Pre-Depends: php-common (>= 2:69~)                                                                                                                           
Depends: phpapi-20180731 | phpapi-20170718 | phpapi-20160303 | phpapi-20151012 | phpapi-20131226, libc6 (>= 2.17)                                            
Recommends: php-apcu-bc                                                                                                                                      
Suggests: php-gd                                                                                                                                             
Conflicts: php-user-cache, php-xcache, php-yac
Replaces: php-user-cache
Homepage: https://pecl.php.net/package/APCu
Download-Size: 86.1 kB
APT-Manual-Installed: yes
APT-Sources: https://packages.sury.org/php stretch/main arm64 Packages
Description: APC User Cache for PHP
 The APCu is userland caching: APC (Alternative PHP Cache) stripped of
 opcode caching after the deployment of Zend OpCache in PHP 5.5 as the
 primary solution to opcode caching in future versions of PHP.
 .
 The APCu is a fast solution for userland caching (and dumping) of PHP
 variables locally, it is not distributed like MemcacheD, but they can
 be used together for optimal caching.

N: There is 1 additional record. Please use the '-a' switch to see it
root@DietPi:~# dpkg -l | grep apcu
ii  php-apcu                    5.1.17+4.0.11-1+0~20190217111312.9+stretch~1.gbp192528 arm64        APC User Cache for PHP
```

**Comment by Fourdee** [comment]
@MichaIng 

Ah error on my part, ignore. Had the APT pinning file still in place whoopsie!
```
root@DietPi:~# apt-get install php7.3-apcu
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Note, selecting 'php-apcu' instead of 'php7.3-apcu'
php-apcu is already the newest version (5.1.17+4.0.11-1+0~20190217111312.9+stretch~1.gbp192528).
0 upgraded, 0 newly installed, 0 to remove and 1 not upgraded.
```

**Comment by MichaIng** [comment]
@Fourdee 
Ah okay 😄. Just verified that ARMv8 APT dependencies are the same, and for all I went through it is:
- All packages of the sury.org repo just require libssl1.1 (>= 1.1.0) which is provided by Stretch already.
- Only the openssl package requires libssl1.1 (>= 1.1.1), but again nothing requires this openssl version.
- So the following APT pinning should work reliable on all systems with sury.org:
```
Package: openssl libssl*
Pin: origin packages.sury.org
Pin-Priority: -1'
```

Only on ARMv6 this is a pain with the Buster repo... However most urgent things resolved with:
```
# Allow to install PHP7.3 dependencies and meta packages from Buster
# - Lighttpd must be pulled as well from Buster since the Stretch version does not support Buster libssl1.1 (1.1.1)
Package: php libapache2-mod-php php-* libssl* libc6* libc-* libgssapi-krb5-2 libpcre2-8-0 libk5crypto3 libkrb5-3 libkrb5support0 locales libzip4 curl lighttpd
Pin: release n=buster
Pin-Priority: 501

# Pin down all other Buster packages to only allow upgrades of already installed ones via: "apt upgrade"
Package: *
Pin: release n=buster
Pin-Priority: 100
```
- I went the way to allow every single package from Buster that is absolutely required.
- `apt install -t buster php7.3-...` leads to ALL dependency packages are pulled from Buster, regardless if the ones from Stretch would be sufficient or not. And there is simply no way to tell APT to respect APT pinning but pull individual packages from Buster, when required to resolve version dependencies. External solvers seem to be able to do this, but those are veeeery slow and I couldn't get it to work either...
- So to prevent running into further distro mix, which causes only more and more dependency issues, or you end up with a pure Buster systems finally, no other way than allowing every single required package manually...

One reason I created the Raspbian Buster image. The faster we get Raspbian Buster stable, the faster we can safe ARMv6 (RPi1+Zero) users from ugly distro version mix 😉.

**Comment by Fourdee** [comment]
Note to self:
- :u6307: Add presets?

**Comment by Fourdee** [comment]
@MichaIng 

Ready for dev/beta testing. I have been testing on my NAS system, seems to work a charm :smiley: 
image
image

**Comment by MichaIng** [comment]
@Fourdee 
Impressive one. Sorry that I could not have a look yet. Will play around and merge at latest tomorrow.

Btw about sysvinit support:
- For now we could simply hide the process tool part for, then slowly (when touching the install code anyway) migrate sysvinit service of remaining software titles to systemd.
- But probably this is how you handle it already 😉.

About `Mode`:
- Probably "Disabled" looks a bid confusing for users which are not aware of systemd and general service handling.
- Perhaps simply show "DietPi controlled (recommended)" and "systemd controlled" instead?

About `Status`:
- We could print the "Active" line directly to the menu, which contains the most relevant info:
```
root@VM-Stretch:~# systemctl -l status cron | grep '^[[:blank:]]*Active:'
   Active: active (running) since Mon 2019-05-20 21:31:28 CEST; 1h 9min ago
```
- Although it is a bid long... Or add this to the top of the menu? Although also doubles a bid with `State` 🤔.

**Comment by Fourdee** [comment]
@MichaIng 

>  migrate sysvinit service of remaining software titles to systemd.

Yep, wrapper created on demand only for sysinit services we have left over, plan to replace all with native systemd units in future.

Regardless, it works.

> Perhaps simply show "DietPi controlled (recommended)" and "systemd controlled" instead?

:+1: Will make the change :u6307: done

> Although it is a bid long... Or add this to the top of the menu? Although also doubles a bid with 

Maybe we could show uptime on its own (eg: 1h 5m ago)? Will take a look.

Cant find any native (since 1hour ago), only start time.
```
root@DietPi:~# systemctl show cron.service | grep ExecMainStartTimestamp=
ExecMainStartTimestamp=Tue 2019-05-21 14:17:30 BST
```

**Comment by Fourdee** [comment]
Ok ready. No further changes my end until dev/beta testing.

**Comment by Fourdee** [comment]
@MichaIng 

Will merge this, its ready. We can tweak/change minor items further if needed on fly.
